src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,530 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+}
+
 namespace {
-TextEncoder::EncodeIntoResult encodeIntoImpl(
-    jsg::Lock& js, jsg::JsString input, jsg::BufferSource& buffer) {
-  auto result = input.writeInto(
-      js, buffer.asArrayPtr().asChars(), jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
-  return TextEncoder::EncodeIntoResult{
-    .read = static_cast<int>(result.read),
-    .written = static_cast<int>(result.written),
-  };
+
+// Find how many Latin-1 characters fit when converted to UTF-8
+// Uses chunked forward scan with SIMD, O(result) complexity
+// Template parameter ReturnLength controls whether to return just position or (position, utf8_length)
+template <bool ReturnLength = false>
+std::conditional_t<ReturnLength, std::pair<size_t, size_t>, size_t> findBestFitLatin1(
+    const char* data, size_t length, size_t bufferSize) {
+  size_t pos = 0;
+  size_t utf8Accumulated = 0;
+  constexpr size_t CHUNK = 256;
+
+  while (pos < length) {
+    size_t remaining = length - pos;
+    size_t chunkSize = kj::min(remaining, CHUNK);
+    size_t chunkUtf8Len = simdutf::utf8_length_from_latin1(data + pos, chunkSize);
+
+    if (utf8Accumulated + chunkUtf8Len > bufferSize) {
+      // Chunk would overflow - binary search within chunk
+      size_t left = 0;
+      size_t right = chunkSize;
+      size_t bestFit = 0;
+
+      while (left <= right) {
+        size_t mid = left + (right - left) / 2;
+        if (mid == 0) break;
+
+        size_t midUtf8Length = simdutf::utf8_length_from_latin1(data + pos, mid);
+        if (utf8Accumulated + midUtf8Length <= bufferSize) {
+          bestFit = mid;
+          left = mid + 1;
+        } else {
+          right = mid - 1;
+        }
+      }
+
+      if constexpr (ReturnLength) {
+        size_t finalPos = pos + bestFit;
+        size_t finalUtf8Len =
+            utf8Accumulated + simdutf::utf8_length_from_latin1(data + pos, bestFit);
+        return {finalPos, finalUtf8Len};
+      } else {
+        return pos + bestFit;
+      }
+    }
+
+    utf8Accumulated += chunkUtf8Len;
+    pos += chunkSize;
+  }
+
+  if constexpr (ReturnLength) {
+    return {pos, utf8Accumulated};
+  } else {
+    return pos;
+  }
 }
-}  // namespace
 
-jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+// Find how many UTF-16 code units fit when converted to UTF-8
+// Uses chunked forward scan with SIMD, O(result) complexity. Never splits surrogate pairs.
+// Template parameter ReturnLength controls whether to return just position or (position, utf8_length)
+template <bool ReturnLength = false>
+std::conditional_t<ReturnLength, std::pair<size_t, size_t>, size_t> findBestFitUtf16(
```

Comment:
These two find best-fit functions, are I think, the biggest source of complexity in this PR. Is there any way to split them off into another PR?

Reply:
I've rewritten these, please take another look.